### PR TITLE
Bin--Leaves constraints

### PR DESCRIPTION
**Pull request by maxTarlov**
Generalized leaf-counting binarity constraints and added updated constraints to the interface.

I had to mess with getDescendantsOfCat (helper function in binarity.js) since it only actually counted terminal nodes of the specified category. Details in commit history and comment within code.

Tests look good, constraints are assigning the correct number of violations now.

### CODE REVIEW

File: constraints/binarity.js [hunk 1]
```diff
@@ -79,35 +79,69 @@ function binMaxBranchesGradient(s, ptree, cat){
 
 /*TRUCKENBRODT-STYLE BINARITY*/
 
-//Parent-category-neutral version of:
-//Sandalo & Truckenbrodt 2002: "Max-Bin: P-phrases consist of maximally two prosodic words"
-//Assigns a violation for every node in ptree that dominates more than two prosodic words.
```
Comment (jbellik) [inline constraints/binarity.js]: We should keep the part of the old comment about Sandalo & Truckenbrodt 2002 back in; the source of the original constraint proposal is useful information.

File: constraints/binarity.js [hunk 2]
```diff
@@ -79,35 +79,69 @@ function binMaxBranchesGradient(s, ptree, cat){
...
 		}
 	}
 	return vcount;
 }
 
-//Gradient version of Truckenbrodt's Maximum Binarity
-function binMax2WordsGradient(s, ptree, cat){
+/* Gradiant BinMax (Leaves)
+* I don't know how to define this constraint in pros, but its binMaxLeaves as
+* a gradiant constraint instead of a catigorical constraint.
```
Comment (jbellik) [inline constraints/binarity.js]: gradient
it's (it is)
categorical

File: constraints/binarity.js [hunk 3]
```diff
@@ -79,35 +79,69 @@ function binMaxBranchesGradient(s, ptree, cat){
...
+*/
+function binMaxLeavesGradient(s, ptree, c){
 	var vcount = 0;
+	//the category we are looking for:
+	var target = pCat.nextLower(c);
+	//pCat.nextLower defined in prosdic hierarchy.js
```
Comment (jbellik) [inline constraints/binarity.js]: prosodicHierarchy.js

File: constraints/binarity.js [hunk 4]
```diff
@@ -122,19 +156,64 @@ function getDescendentsOfCat(x, cat){
...
+			}
 			var yDescendents = getDescendentsOfCat(x.children[y], cat);
 			for(var i=0; i < yDescendents.length; i++){
 				descendents.push(yDescendents[i]);
 			}
 		}
 	}
+	/* this else if statement was double counting terminal nodes of category cat.
+	 * removing it causes double counting to stop, but if the input is a terminal
+	 * node, function will return 0.
```
Comment (jbellik) [inline constraints/binarity.js]: A terminal node has no descendants, so returning 0 is fine.

File: constraints/binarity.js (hunk 1 above)
Comment (maxTarlov) [reply constraints/binarity.js]: comment restored in latest commit starting on line 87

**Comment by maxTarlov** [comment]
I fixed the various spelling errors and omissions in the latest commit.

**Pull request merged**